### DISCUSSION TITLE
Non-blocking buffer / geometry GPU upload

### DISCUSSION THREAD

**Discussion by christjt** [gpuweb/gpuweb#3079] [Q&A] [answered]
Hello 👋 

I was wondering if there is any way one could do async / non-blocking uploading of geometry to the GPU. 

Hypothetical case:
Have some app running with some simple geometry being visualized and based on some user action the application fetches a really complex triangle mesh to be visualized. Doing this the trivial way would cause major blocking while the buffer is being sent to the GPU (or if I issue a draw, it will wait I presume). 

Conceptually, my current approach would be to send the data to some staging buffer and then have a fence that signals when it is done, and only then start issuing draws / starting executing the relevant pipeline (assuming I also blit the data from staging to the internal memory). Am I totally off base here or is there some other preferred way of not dropping frames when loading new data? This is really cumbersome in WebGL and have some hope this can be done using WebGPU. For my case I don't care about waiting for the data being ready, i.e. just don't include it in the render loop before the data is ready to be drawn.

From what I can tell `write_buffer` seems to be not blocking itself, but if I start executing a pipeline that depends on this data I assume it will wait / block the frame.

**Comment by Kangz** [comment, marked as answer]
At the moment, none of the WebGPU operations are blocking, by design, to avoid stalling the Web page. Depending on the amount of data `GPUQueue.writeBuffer` may "stall" as it has to immediately `memcpy` the data somewhere. However none of the further operations happening on the GPU are "stalled", they are just pipeline until after the `writeBuffer` is complete. No need to wait for a fence or anything.

If you are worried about the stalling in `writeBuffer` due to the `memcpy`, you can instead use a mappable buffer that you write into, so that you can control how much copying happens at which times on the CPU, to avoid stalls. Either a `createBuffer({..., mappedAtCreation: true})` or `mapAsync` would work for that. If you're also worried about the cost of copying from the CPU/GPU-visible to private GPU memory, you could schedule copies over multiple frames if needed. 

Overall though, I'd suggest you use the simplest solution that works for you right now, and measure whether there's a stall or not. Uploading tens of megabytes in a frame should be perfectly reasonable so you might not need to do any optimizations.

**Reply by christjt** [reply]
Awesome thanks. So from what I get here nothing (other than maybe the memcpy) will stall the CPU. But if I writeBuffer -> execute some pipeline that uses the buffer it will have to wait for it to render the frame, right? In any case, thanks for the remarkably fast response, and I will profile the different solutions.
